plugin/plugin.go

```diff
@@ -382,23 +406,151 @@ func (p *AzurePlugin) ListHosts(ctx context.Context, req *pb.ListHostsRequest) (
+					}
+					if props.NetworkProfile == nil {
+						return nil, fmt.Errorf("error fetching network profile for vm with id %q", *res.ID)
+					}
+
+					// Within the VM, catalog the various interfaces
+					for _, ifaceRef := range *props.NetworkProfile.NetworkInterfaces {
+						if ifaceRef.ID == nil {
+							return nil, fmt.Errorf("nil ID for network interface for vm with id %q", *res.ID)
+						}
+						ifName := extractResourceSuffix(*ifaceRef.ID)
+						iface, err := ifClient.GetVirtualMachineScaleSetNetworkInterface(ctx, resourceGroup, name, *vmssvm.InstanceID, ifName, "")
+						if err != nil {
+							return nil, fmt.Errorf("error fetching network interface with id %q: %w", *ifaceRef.ID, err)
+						}
+						if iface.InterfacePropertiesFormat == nil || iface.InterfacePropertiesFormat.IPConfigurations == nil {
+							continue
+						}
+						var netInfo networkInfo
+						for _, ipconf := range *iface.InterfacePropertiesFormat.IPConfigurations {
+							if ipconf.PrivateIPAddress != nil {
+								netInfo.IpAddresses = append(netInfo.IpAddresses, *ipconf.PrivateIPAddress)
+							}
+							if ipconf.PublicIPAddress != nil {
+								if ipconf.PublicIPAddress.ID == nil {
+									return nil, fmt.Errorf("ip configuration %q has public IP address info but nil id", *ipconf.ID)
+								}
+								//ipResGroup, ipName, err := splitId(*ipconf.PublicIPAddress.ID, constMsNetworkService, constPublicIpAddressesResource)
+								if err != nil {
+									return nil, fmt.Errorf("error splitting public ip address id %q: %w", *ipconf.PublicIPAddress.ID, err)
+								}
+								pipName := extractResourceSuffix(*ipconf.PublicIPAddress.ID)
+								pipInfo, err := pipClient.GetVirtualMachineScaleSetPublicIPAddress(ctx, resourceGroup, *vmss.Name, *vmssvm.InstanceID, ifName, *ipconf.Name, pipName, "")
+								if err != nil {
+									if err != nil {
+										return nil, fmt.Errorf("error fetching public IP information with resource group %q and name %q: %w", resourceGroup, pipName, err)
+									}
+								}
+								if pipInfo.PublicIPAddressPropertiesFormat == nil {
+									return nil, fmt.Errorf("nil public ip address properties format for public ip %q", *ipconf.PublicIPAddress.ID)
+								}
+								if pipInfo.PublicIPAddressPropertiesFormat.IPAddress != nil {
+									netInfo.IpAddresses = append(netInfo.IpAddresses, *pipInfo.PublicIPAddressPropertiesFormat.IPAddress)
+								}
+							}
+						}
+						vmToNetworkMap[*vmssvm.ID] = netInfo
+					}
+
+				}
 			}
-			vmToNetworkMap[vmId] = netInfo
 		}
 	}
 
 	ret := &pb.ListHostsResponse{}
 	for k, v := range vmToNetworkMap {
-		_, externalName, err := splitId(k, constMsComputeService, constVirtualMachinesResource)
-		if err != nil {
-			externalName = ""
+		var host *pb.ListHostsResponseHost
+		// see if we're dealing with a VMSS instance or a VM
+		if strings.Contains(k, constMsComputeService+"/"+constVirtualMachineScaleSetsResource) {
```

Comment:
Is `Contains` reliable enough here? Does it not have to be an exact match or at least a prefix match? What values can `k` take?